includes/Story_Archive.php

```diff
@@ -0,0 +1,195 @@
+	 *
+	 * @var Settings Settings instance.
+	 */
+	private $settings;
+
+	/**
+	 * Story_Post_Type instance.
+	 *
+	 * @var Story_Post_Type Story_Post_Type instance.
+	 */
+	private $story_post_type;
+
+	/**
+	 * Analytics constructor.
+	 *
+	 * @since 1.13.0
+	 *
+	 * @param Settings        $settings        Settings instance.
+	 * @param Story_Post_Type $story_post_type Experiments instance.
+	 *
+	 * @return void
+	 */
+	public function __construct( Settings $settings, Story_Post_Type $story_post_type ) {
+		$this->settings        = $settings;
+		$this->story_post_type = $story_post_type;
+	}
+
+	/**
+	 * Registers Filters and actions
+	 *
+	 * @since 1.13.0
+	 *
+	 * @return void
+	 */
+	public function register() {
+		add_filter( 'pre_handle_404', [ $this, 'redirect_post_type_archive_urls' ], 10, 2 );
+
+		add_action( 'add_option_' . $this->settings::SETTING_NAME_ARCHIVE, [ $this, 'update_archive_setting' ] );
+		add_action( 'update_option_' . $this->settings::SETTING_NAME_ARCHIVE, [ $this, 'update_archive_setting' ] );
+		add_action( 'add_option_' . $this->settings::SETTING_NAME_ARCHIVE_PAGE_ID, [ $this, 'update_archive_setting' ] );
+		add_action( 'update_option_' . $this->settings::SETTING_NAME_ARCHIVE_PAGE_ID, [ $this, 'update_archive_setting' ] );
+
+		add_filter( 'display_post_states', [ $this, 'filter_display_post_states' ], 10, 2 );
+		add_action( 'pre_get_posts', [ $this, 'pre_get_posts' ] );
+	}
+
+	/**
+	 * Handles redirects to the post type archive.
+	 *
+	 * @since 1.13.0
+	 *
+	 * @param bool|mixed $bypass Pass-through of the pre_handle_404 filter value.
+	 * @param \WP_Query  $query  The WP_Query object.
+	 *
+	 * @return bool|mixed Whether to pass-through or not.
+	 */
+	public function redirect_post_type_archive_urls( $bypass, $query ) {
+		global $wp_rewrite;
+
+		if ( $bypass || ! is_string( $this->story_post_type->get_has_archive() ) || ( ! $wp_rewrite instanceof WP_Rewrite || ! $wp_rewrite->using_permalinks() ) ) {
+			return $bypass;
+		}
+
+		// 'pagename' is for most permalink types, name is for when the %postname% is used as a top-level field.
+		if ( $this->story_post_type::REWRITE_SLUG === $query->get( 'pagename' ) || $this->story_post_type::REWRITE_SLUG === $query->get( 'name' ) ) {
+			$redirect_url = get_post_type_archive_link( $this->story_post_type::POST_TYPE_SLUG );
+
+			if ( ! $redirect_url ) {
+				return $bypass;
+			}
+
+			// Only exit if there was actually a location to redirect to.
+			// Allows filtering location in tests to verify behavior.
+			if ( wp_safe_redirect( $redirect_url, 301 ) ) {
+				exit;
+			}
+		}
+
+		return $bypass;
+	}
+
+	/**
+	 * Clear rewrite rules on update on setting.
+	 *
+	 * @since 1.13.0
+	 *
+	 * @return void
+	 */
+	public function update_archive_setting() {
+		if ( ! defined( '\WPCOM_IS_VIP_ENV' ) || false === \WPCOM_IS_VIP_ENV ) {
+			$this->story_post_type->unregister_post_type();
+			$this->story_post_type->register_post_type();
+			flush_rewrite_rules( false );
+		}
```

Comment:
I wonder if this shouldn't have been like this all along 🤔 

```suggestion
		$this->story_post_type->unregister_post_type();
		$this->story_post_type->register_post_type();
			
		if ( ! defined( '\WPCOM_IS_VIP_ENV' ) || false === \WPCOM_IS_VIP_ENV ) {
			flush_rewrite_rules( false );
		}
```

Reply:
Why not.